### PR TITLE
Cache docker builds

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/3637#discussion_r494210629

Used `docker/build-push-action@v2` to able to cache

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ydcjeff** [comment]
@Borda I used [`docker/build-push-action@v2`](https://github.com/docker/build-push-action) to cache docker builds (can cache locally or on registry).
And we only cache `base-cuda` and `base-xla` builds in nightly events, since
caching in `base-conda` took more time than no cache, so no cache right now.

Also, storing cache locally took more time when restoring the cache.
I have tested in the forked repo and restoring local cache took ~3 mins while
restoring from registry cache took ~2s, idk why.

So, does pushing cache images (separate images) to docker hub sound good to you?

File: .github/workflows/nightly.yml
Comment (Borda) [inline .github/workflows/nightly.yml:120]: lets keep this in single run, sowe do not think twire if cuda and conda is paired

Comment (ydcjeff) [reply .github/workflows/nightly.yml:120]: Yes, done!

**Comment by ydcjeff** [comment]
@Borda  Shall we install nvidia-dali in this PR or in a later PR?

File: .github/workflows/nightly.yml
Comment (Borda) [reply .github/workflows/nightly.yml:120]: well, I still see it here :]

**Comment by Borda** [comment]
> @Borda Shall we install nvidia-dali in this PR or in a later PR?

I would keep it paired with the adding DALI example...

**Comment by edenlightning** [comment]
hi @ydcjeff mind taking a look at failing tests?

File: .github/workflows/ci_dockers.yml
Comment (Borda) [inline .github/workflows/ci_dockers.yml:23]: I was checking this but not sure why it is needed, mind add a comment here...

**Review by Borda — APPROVED** [review]
lgtm, just wait till all CI passes

File: .github/workflows/ci_dockers.yml
Comment (ydcjeff) [reply .github/workflows/ci_dockers.yml:23]: `build-push-action@v2` has `cache-from` and `cache-to` arguments which come from `buildx` command.
https://github.com/docker/buildx#buildx-build-options-path--url---
so we need `setup-buildx-action`

will add a comment.

**Comment by ydcjeff** [comment]
@Borda  Before merging, just want to make sure

this PR stores the cache metadata of the docker builds on docker hub with the separate names.
I found caching on docker hub is faster than caching in CI
Does it sounds good to you?

**Comment by Borda** [comment]
> this PR stores the cache metadata of the docker builds on docker hub with the separate names.

how does it appear on Dockerhub, as a new entry/tag?

**Comment by ydcjeff** [comment]
> > this PR stores the cache metadata of the docker builds on docker hub with the separate names.
> 
> how does it appear on Dockerhub, as a new entry/tag?

It will appear as a new tag like `base-cuda-cache-py3.8-torch1.6` (for eg), DIGEST would be some hashes (I think), OS/ARCH will be none, there will be no compressed size.

File: .github/workflows/nightly.yml
Comment (ydcjeff) [inline .github/workflows/nightly.yml:116]: This will appear as a new tag on docker hub

File: .github/workflows/ci_dockers.yml
Comment (tchaton) [inline .github/workflows/ci_dockers.yml:28]: comment not related anymore

**Review by tchaton — APPROVED** [review]
Great addition ! I didn't know about `cache-from` and `cache-to`. Really lean API from `docker/build-push-action@v2` :)

File: .github/workflows/ci_dockers.yml
Comment (ydcjeff) [reply .github/workflows/ci_dockers.yml:28]: you mean `publish master` comment?

**Review by awaelchli — APPROVED** [review]

**Comment by ydcjeff** [comment]
I know 'Ready to go' is not meant to be added by PR author. But `fix build-args` commit changes small, so I think it's good to go.

**Pull request merged**
